src/react/utils.js

```diff
@@ -58,6 +60,95 @@ export function valueIsClassComponent(realm: Realm, value: Value) {
   return false;
 }
 
+// a nested object of a React Element should be hoisted where all its properties are known
+// at evaluation time to be "static" or "pure". This means function values in scope, strings, numbers
+// and deeply nested object/arrays that share the same heuristics
+function canHoistObject(realm: Realm, object: ObjectValue, hoistedScope?: Generator): boolean {
+  if (isReactElement(object)) {
+    return canHoistReactElement(realm, object);
+  }
+  for (let [propName] of object.properties) {
+    let prop = Get(realm, object, propName);
+    if (prop instanceof AbstractValue) {
+      return false;
+    } else if (prop instanceof FunctionValue && prop.parent !== hoistedScope) {
+      return false;
+    } else if (prop instanceof ArrayValue) {
+      let isHoistable = canHoistArray(realm, prop, hoistedScope);
+
+      if (!isHoistable) {
+        return false;
+      }
+    } else if (prop instanceof ObjectValue) {
+      let isHoistable = canHoistObject(realm, prop, hoistedScope);
+
+      if (!isHoistable) {
+        return false;
+      }
+    }
+  }
+  return true;
+}
+
+function canHoistArray(realm: Realm, array: ArrayValue, hoistedScope?: Generator): boolean {
+  let lengthValue = Get(realm, array, "length");
+  invariant(lengthValue instanceof NumberValue);
+  let length = lengthValue.value;
+  for (let i = 0; i < length; i++) {
+    let element = Get(realm, array, "" + i);
+
+    if (element instanceof AbstractValue) {
+      return false;
+    } else if (element instanceof ArrayValue) {
+      let isHoistable = canHoistArray(realm, element, hoistedScope);
+
+      if (!isHoistable) {
+        return false;
+      }
+    } else if (element instanceof FunctionValue && element.parent !== hoistedScope) {
+      return false;
+    } else if (element instanceof ObjectValue) {
+      let isHoistable = canHoistObject(realm, element, hoistedScope);
+
+      if (!isHoistable) {
+        return false;
+      }
+    }
+  }
+  return true;
+}
+
+export function canHoistReactElement(realm: Realm, reactElement: ObjectValue, hoistedScope?: Generator): boolean {
+  if (reactElement.$CanHoist) {
+    return true;
+  }
+  // we host components <div />, <span /> etc
+  // we also allow functions where their parent scope matches the hoisted scope
+  let type = Get(realm, reactElement, "type");
+  if (!(type instanceof StringValue || (type instanceof FunctionValue && type.parent === hoistedScope))) {
```

Comment:
I was thinking that we should be able to hoist newly created functions too, assume that the function doesn't close over abstract values, like any other object.

```js
function Foo() {
  var Bar = () => <div />;
  return <Bar />;
}
```
->
```js
var Bar = () => <div />;
var _bar = <Bar />;

function Foo() {
  return _bar;
}
```

But then I realized that this is not equivalent since we use the identity of `type` to determine if state is preserved and it would be a new one every time.

The same quirk applies to other objects too though. So maybe we should just live with that?

Reply:
Your above example doesn't really matter here. It will be evaluated by the reconciler and inlined before it gets to the hoisting logic. There are already tests that show this. :)

Reply:
It matters if it is passed to something that doesn't get inlined though.

```js
function Foo() {
  var Bar = () => <div />;
  return <NotInlined><Bar /></NotInlined>;
}
```

Reply:
It handled closures now. So this shouldn't be an issue :)